File: RaftSwift/Sources/RaftGRPCTransport/Utils/ServerIDInjectionInterceptor.swift
import GRPCCore

/// Injects the peer ID into the request metadata
struct ServerIDInjectionInterceptor: ClientInterceptor {
    /// The ID of the peer sending the requests
    let peerID: Int

    func intercept<Input, Output>(
        request: StreamingClientRequest<Input>,
        context: ClientContext,
        next: (
            _ request: StreamingClientRequest<Input>,
            _ context: ClientContext
        ) async throws -> StreamingClientResponse<Output>
    ) async throws -> StreamingClientResponse<Output>
        where Input: Sendable, Output: Sendable
    {
        var request = request
        request.metadata.addString(String(peerID), forKey: "x-peer-id")
        return try await next(request, context)
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/transport/grpc/ServerIDInjectionInterceptor.kt
package raft.transport.grpc

import io.grpc.*

/**
 * Injects the peer ID into the request metadata.
 */
class ServerIDInjectionInterceptor(
    /**
     * The ID of the peer sending the requests.
     */
    private val peerID: Int
) : ClientInterceptor {

    override fun <ReqT, RespT> interceptCall(
        method: MethodDescriptor<ReqT, RespT>,
        callOptions: CallOptions,
        next: Channel
    ): ClientCall<ReqT, RespT> {
        return object : ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
            next.newCall(method, callOptions)
        ) {
            override fun start(responseListener: Listener<RespT>, headers: Metadata) {
                headers.put(
                    Metadata.Key.of("x-peer-id", Metadata.ASCII_STRING_MARSHALLER),
                    peerID.toString()
                )
                super.start(responseListener, headers)
            }
        }
    }
}